### PR TITLE
tools/rdmaucma: Add new command

### PR DESCRIPTION
**Pull request by pizhenwei**
Add rdmaucma tools to trace RDMA UCMA events.
This can be useful to analyze issues on RDMA CM.

### CODE REVIEW

File: tools/rdmaucma.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        "device removal",   \
+        "multicast join",   \
+        "multicast error",  \
+        "address change",   \
+        "timewait exit"     ]
+
+def print_ipv4_event(cpu, data, size):
+    event = b["ipv4_events"].event(data)
+
+    cm_event = "unkonw event"
```
Comment (chenhengqi) [inline tools/rdmaucma.py:135]: typo: unknown

File: tools/rdmaucma.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    if args.ebpf:
+        exit()
+
+# load BPF program
+b = BPF(text=bpf_text)
+b.attach_kprobe(event="ucma_event_handler", fn_name="trace_ucma_event_handler")
+
+# see linux/include/rdma/rdma_cm.h
+rdma_cm_event = [           \
+        "address resolved", \
```
Comment (chenhengqi) [inline tools/rdmaucma.py:115]: No need for `\`.

File: tools/rdmaucma_example.txt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+# ./rdmaucma
+Tracing RDMA Userspace Connection Manager Access event... Hit Ctrl-C to end.
+Timestamp Event            Family Local                     Remote
+16:59:09  connect request  IPv6   fbcc:da4a:15:479::165:6379  <-> fbcc:da4a:15:479::166:57012
+16:59:09  established      IPv6   fbcc:da4a:15:479::165:6379  <-> fbcc:da4a:15:479::166:57012
+16:59:11  disconnected     IPv6   fbcc:da4a:15:479::165:6379  <-> fbcc:da4a:15:479::166:57012
+16:59:13  connect request  IPv6   fbcc:da4a:15:479::165:6379  <-> fbcc:da4a:15:479::166:36899
+16:59:13  established      IPv6   fbcc:da4a:15:479::165:6379  <-> fbcc:da4a:15:479::166:36899
+16:59:14  disconnected     IPv6   fbcc:da4a:15:479::165:6379  <-> fbcc:da4a:15:479::166:36899
+16:59:23  connect request  IPv4    192.168.122.165:6379  <->  192.168.122.166:21119
```
Comment (chenhengqi) [inline tools/rdmaucma_example.txt:17]: The output is not aligned.

File: tools/rdmaucma.py (hunk 1 above)
Comment (pizhenwei) [reply tools/rdmaucma.py]: OK.

File: tools/rdmaucma.py (hunk 2 above)
Comment (pizhenwei) [reply tools/rdmaucma.py]: OK.

File: tools/rdmaucma_example.txt (hunk 3 above)
Comment (pizhenwei) [reply tools/rdmaucma_example.txt]: OK.

**Comment by pizhenwei** [comment]
@chenhengqi 
Force pushed a new version(34bdcc9827582d2c669e52cced947ac9c2bd5aed). Fix:
* typo
* remove unnecessary '\'
* refine output

**Comment by pizhenwei** [comment]
@yonghong-song @chenhengqi 
PING!

**Comment by yonghong-song** [comment]
I have to say that I am not very familiar with RDMA so it is hard for me to judge how useful this tool could be. If somebody else also has some RDMA experience, maybe comment this patch as well?

**Comment by pizhenwei** [comment]
> I have to say that I am not very familiar with RDMA so it is hard for me to judge how useful this tool could be. If somebody else also has some RDMA experience, maybe comment this patch as well?

Hi, @jgunthorpe
During I debug RDMA programs, I realize that I need a tool to show the UCMA events, so I develop this [tool](https://github.com/iovisor/bcc/pull/4616/files#diff-f994c8708393c0f440a494a345af92596aabee333f34dbc138f87ea734c094cd). 

Do you have any suggestions on this tool?

**Comment by pizhenwei** [comment]
> > > I have to say that I am not very familiar with RDMA so it is hard for me to judge how useful this tool could be. If somebody else also has some RDMA experience, maybe comment this patch as well?
> > 
> > 
> > Hi, @jgunthorpe @rleon During I debug RDMA programs, I realize that I need a tool to show the UCMA events, so I develop this [tool](https://github.com/iovisor/bcc/pull/4616/files#diff-f994c8708393c0f440a494a345af92596aabee333f34dbc138f87ea734c094cd).
> > Do you have any suggestions on this tool?
> 
> I'm not against this tool, just wonder why standard tracing which exists in the kernel is not enough? https://elixir.bootlin.com/linux/v6.4-rc7/source/drivers/infiniband/core/cm_trace.h

This was introduced since linux-v5.10, it is not widely deployed in the production environment.
I'd like to push a new version of this tool: 1, try to use trace point. 2, if 1 fails, try to use kprobe. Do you agree this?

**Comment by rleon** [comment]
> > > > I have to say that I am not very familiar with RDMA so it is hard for me to judge how useful this tool could be. If somebody else also has some RDMA experience, maybe comment this patch as well?
> > > 
> > > 
> > > Hi, @jgunthorpe @rleon During I debug RDMA programs, I realize that I need a tool to show the UCMA events, so I develop this [tool](https://github.com/iovisor/bcc/pull/4616/files#diff-f994c8708393c0f440a494a345af92596aabee333f34dbc138f87ea734c094cd).
> > > Do you have any suggestions on this tool?
> > 
> > 
> > I'm not against this tool, just wonder why standard tracing which exists in the kernel is not enough? https://elixir.bootlin.com/linux/v6.4-rc7/source/drivers/infiniband/core/cm_trace.h
> 
> This was introduced since linux-v5.10, it is not widely deployed in the production environment. I'd like to push a new version of this tool: 1, try to use trace point. 2, if 1 fails, try to use kprobe. Do you agree?

I don't know anything about code expectation in this repository. It is better to ask relevant maintainers.

**Comment by pizhenwei** [comment]
> > > > I have to say that I am not very familiar with RDMA so it is hard for me to judge how useful this tool could be. If somebody else also has some RDMA experience, maybe comment this patch as well?
> > > 
> > > 
> > > Hi, @jgunthorpe @rleon During I debug RDMA programs, I realize that I need a tool to show the UCMA events, so I develop this [tool](https://github.com/iovisor/bcc/pull/4616/files#diff-f994c8708393c0f440a494a345af92596aabee333f34dbc138f87ea734c094cd).
> > > Do you have any suggestions on this tool?
> > 
> > 
> > I'm not against this tool, just wonder why standard tracing which exists in the kernel is not enough? https://elixir.bootlin.com/linux/v6.4-rc7/source/drivers/infiniband/core/cm_trace.h
> 
> This was introduced since linux-v5.10, it is not widely deployed in the production environment. I'd like to push a new version of this tool: 1, try to use trace point. 2, if 1 fails, try to use kprobe. Do you agree?

@yonghong-song 
This version is tested on linux-5.4 and linux-5.10, both work fine. Do you have any concern?

**Review by Thebruddah — APPROVED** [review]
Sync

**Comment by pizhenwei** [comment]
@yonghong-song PING!

**Comment by yonghong-song** [comment]
Okay, I tried on one of our internal hosts, loading rdma_ucm module,  the tool seems working fine.
I will merge this tool thinking it MIGHT be useful although I am not a direct user for it.

As a followup, please add the tool into README.md tools list.

**Pull request merged**